helix-view/src/editor.rs

```diff
@@ -398,18 +412,88 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
+
+    pub fn close_document(&mut self, doc_id: DocumentId, force: bool) -> anyhow::Result<()> {
+        if !self.documents.contains_key(&doc_id) {
+            anyhow::bail!("document does not exist");
+        }
 
-            if let Some(language_server) = doc.language_server() {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        let doc = &self.documents[&doc_id];
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let mut v = vec![];
+
+        for (view, _focused) in self.tree.views_mut() {
+            view.jumps.remove(&doc_id);
+
+            if let Some(alt) = view.last_accessed_doc {
+                if alt == doc_id {
+                    view.last_accessed_doc = None;
+                }
+            }
+
+            if view.doc == doc_id {
+                let new_id = if let Some(alt) = view.last_accessed_doc {
```

Comment:
> For the most part, I've taken inspiration from how nvim handles splits (at least, in my neovim with its various plugins.....) -- namely, that closing a buffer that's visible in multiple splits will also close those splits. 

This actually doesn't seem to do that but instead tries to display a different document in the split? Might be a bit better actually since it keeps the layout intact.

If you'd like to keep the vim behaviour on the other hand, then this code can be simplified: simply `self.close` all the views where `view.doc == doc_id`, then remove the document from `self.documents`. If you're left with an empty view tree but `!self.documents.is_empty()`, then push a new view and set it to `self.documents.iter().first()`

Reply:
The `last_accessed_doc` was so that we could switch to that doc if it was `Some(...)`, since it's more likely that the user wants to see the document they were looking at before closing their current one, than it is that they want to see the first document...

By unconditionally closing the view, we can't check the `last_accessed_doc` and switch to that if it exists... So opening `ab`, `cd`, and `ef`, then closing `ef` will bring you to `ab` instead of (as previously) `cd`. This seems weird to me.